lib/cisco_node_utils/pim.rb

```diff
@@ -0,0 +1,92 @@
+# PIM feature
+# Provides configuration of PIM and its various properties
+# like ssm-range, etc
+#
+# Smitha Gopalan, November 2015
+#
+# Copyright (c) 2014-2015 Cisco and/or its affiliates.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#-------------------------------------------------------
+# CLI: ip pim ssm-range <range>  (under different VRFs)
+#-------------------------------------------------------
+
+require_relative 'node_util'
+
+module Cisco
+  # node_utils class for Pim
+  class Pim < NodeUtil
+    attr_reader :vrf
+
+    # Constructor with vrf
+    # ---------------------
+    def initialize(vrf='default', instantiate=true)
+      fail ArgumentError unless vrf.is_a? String
+      fail ArgumentError unless vrf.length > 0
+      @vrf = vrf
+      if @vrf == 'default'
+        @get_args = @set_args = { state: '' }
+      else
+        @get_args = @set_args = { state: '', vrf: @vrf }
+      end
+      enable if instantiate && !Pim.enabled
+    end
```

Comment:
For v4/v6 you'll need something like this:

*(I simplified the initialize a little)*
```
    def initialize(afi, vrf, instantiate=true)
      fail ArgumentError unless vrf.is_a?(String) && (vrf.length > 0)
      keys = { afi: Pim.afi_cli(afi)}
      keys[:vrf] = vrf unless vrf == 'default'
      @set_args = @get_args = keys
      enable if instantiate
    end

    def self.afi_cli(afi)
      fail ArgumentError, "Argument afi must be 'ipv4' or 'ipv6'" unless
        afi[/(ipv4|ipv6)/]
      afi[/ipv4/] ? 'ip' : afi
    end
```